File: src/clients/python/pysharpcoredb/src/pysharpcoredb/pool.py
"""Connection pooling for PySharpDB."""

import asyncio
import logging
import time
from typing import Dict, List, Optional
from contextlib import asynccontextmanager

from .connection import Connection
from .exceptions import ConnectionError

logger = logging.getLogger(__name__)


class PooledConnection:
    """A pooled database connection."""

    def __init__(self, connection: Connection, pool: 'ConnectionPool'):
        self.connection = connection
        self.pool = pool
        self.created_at = time.time()
        self.last_used = time.time()
        self._closed = False

    async def close(self):
        """Return connection to pool."""
        if not self._closed:
            self._closed = False
            await self.pool._return_connection(self)

    async def __aenter__(self):
        return self.connection

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()

    @property
    def is_closed(self) -> bool:
        return self._closed


class ConnectionPool:
    """Connection pool for PySharpDB.

    Provides efficient connection reuse and management for high-concurrency applications.
    """

    def __init__(
        self,
        host: str,
        port: Optional[int] = None,
        database: str = "default",
        username: Optional[str] = None,
        password: Optional[str] = None,
        tls: bool = True,
        min_connections: int = 1,
        max_connections: int = 10,
        max_idle_time: float = 300.0,  # 5 minutes
        max_lifetime: float = 3600.0,  # 1 hour
        acquire_timeout: float = 30.0
    ):
        """Initialize the connection pool.

        Args:
            host: Server hostname
            port: Server port
            database: Database name
            username: Username
            password: Password
            tls: Use TLS
            min_connections: Minimum connections to maintain
            max_connections: Maximum connections allowed
            max_idle_time: Max idle time before closing connection
            max_lifetime: Max lifetime before recycling connection
            acquire_timeout: Timeout for acquiring connection
        """
        self.host = host
        self.port = port
        self.database = database
        self.username = username
        self.password = password
        self.tls = tls

        self.min_connections = min_connections
        self.max_connections = max_connections
        self.max_idle_time = max_idle_time
        self.max_lifetime = max_lifetime
        self.acquire_timeout = acquire_timeout

        self._available: List[PooledConnection] = []
        self._in_use: Dict[Connection, PooledConnection] = {}
        self._lock = asyncio.Lock()
        self._closed = False

        # Statistics
        self._created_count = 0
        self._destroyed_count = 0

    async def get_connection(self) -> PooledConnection:
        """Get a connection from the pool."""
        if self._closed:
            raise ConnectionError("Connection pool is closed")

        async with self._lock:
            # Try to get an available connection
            while self._available:
                pooled_conn = self._available.pop()
                if self._is_connection_valid(pooled_conn):
                    pooled_conn.last_used = time.time()
                    self._in_use[pooled_conn.connection] = pooled_conn
                    return pooled_conn
                else:
                    await self._destroy_connection(pooled_conn)

            # Create new connection if under limit
            if len(self._in_use) < self.max_connections:
                connection = Connection(
                    host=self.host,
                    port=self.port,
                    database=self.database,
                    username=self.username,
                    password=self.password,
                    tls=self.tls
                )
                await connection.connect()

                pooled_conn = PooledConnection(connection, self)
                self._created_count += 1
                self._in_use[connection] = pooled_conn
                return pooled_conn

            # Wait for a connection to become available
            # For simplicity, we'll just raise an error if at max capacity
            raise ConnectionError(
                f"Connection pool exhausted: {len(self._in_use)}/{self.max_connections} connections in use"
            )

    async def _return_connection(self, pooled_conn: PooledConnection):
        """Return a connection to the pool."""
        async with self._lock:
            if pooled_conn.connection in self._in_use:
                del self._in_use[pooled_conn.connection]

            if self._is_connection_valid(pooled_conn):
                self._available.append(pooled_conn)
            else:
                await self._destroy_connection(pooled_conn)

    async def _destroy_connection(self, pooled_conn: PooledConnection):
        """Destroy a connection."""
        try:
            await pooled_conn.connection.disconnect()
        except Exception as e:
            logger.warning(f"Error closing connection: {e}")

        self._destroyed_count += 1

    def _is_connection_valid(self, pooled_conn: PooledConnection) -> bool:
        """Check if a connection is still valid."""
        now = time.time()

        # Check lifetime
        if now - pooled_conn.created_at > self.max_lifetime:
            return False

        # Check idle time
        if now - pooled_conn.last_used > self.max_idle_time:
            return False

        # Check if connection is still connected
        return pooled_conn.connection.is_connected

    async def close(self):
        """Close the connection pool and destroy all connections."""
        if self._closed:
            return

        async with self._lock:
            self._closed = True

            # Close all available connections
            destroy_tasks = []
            for pooled_conn in self._available:
                destroy_tasks.append(self._destroy_connection(pooled_conn))

            # Close all in-use connections
            for pooled_conn in self._in_use.values():
                destroy_tasks.append(self._destroy_connection(pooled_conn))

            self._available.clear()
            self._in_use.clear()

            await asyncio.gather(*destroy_tasks, return_exceptions=True)

    @property
    def is_closed(self) -> bool:
        """Check if the pool is closed."""
        return self._closed

    @property
    def stats(self) -> Dict[str, int]:
        """Get pool statistics."""
        return {
            "available": len(self._available),
            "in_use": len(self._in_use),
            "total_created": self._created_count,
            "total_destroyed": self._destroyed_count,
            "max_connections": self.max_connections
        }

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()


@asynccontextmanager
async def create_pool(**kwargs):
    """Create a connection pool.

    Args:
        **kwargs: Connection pool parameters

    Yields:
        ConnectionPool instance
    """
    pool = ConnectionPool(**kwargs)
    try:
        yield pool
    finally:
        await pool.close()

File: src/SharpCoreDBServerScriptClients/javascript/sharpcoredb-client/src/pool.ts
/**
 * Connection pooling for @sharpcoredb/client
 */

import { EventEmitter } from 'events';
import { Connection } from './connection';
import { PoolOptions, ConnectionOptions } from './types';
import { ConnectionError } from './errors';

/**
 * A pooled database connection
 */
export class PooledConnection extends EventEmitter {
  private _connection: Connection;
  private _pool: ConnectionPool;
  private _createdAt: number;
  private _lastUsed: number;
  private _closed: boolean;

  constructor(connection: Connection, pool: ConnectionPool) {
    super();
    this._connection = connection;
    this._pool = pool;
    this._createdAt = Date.now();
    this._lastUsed = Date.now();
    this._closed = false;

    // Forward connection events
    this._connection.on('connected', (...args) => this.emit('connected', ...args));
    this._connection.on('disconnected', (...args) => this.emit('disconnected', ...args));
    this._connection.on('error', (...args) => this.emit('error', ...args));
  }

  /**
   * Get the underlying connection
   */
  get connection(): Connection {
    return this._connection;
  }

  /**
   * Check if the pooled connection is closed
   */
  get isClosed(): boolean {
    return this._closed;
  }

  /**
   * Get creation timestamp
   */
  get createdAt(): number {
    return this._createdAt;
  }

  /**
   * Get last used timestamp
   */
  get lastUsed(): number {
    return this._lastUsed;
  }

  /**
   * Update last used timestamp
   */
  markUsed(): void {
    this._lastUsed = Date.now();
  }

  /**
   * Return connection to pool
   */
  async close(): Promise<void> {
    if (this._closed) {
      return;
    }
    this._closed = true;
    await this._pool._returnConnection(this);
  }

  /**
   * Check if connection is valid
   */
  isValid(maxIdleTime: number, maxLifetime: number): boolean {
    const now = Date.now();

    // Check lifetime
    if (now - this._createdAt > maxLifetime) {
      return false;
    }

    // Check idle time
    if (now - this._lastUsed > maxIdleTime) {
      return false;
    }

    // Check if underlying connection is still connected
    return this._connection.isConnected;
  }
}

/**
 * Connection pool for efficient connection reuse
 */
export class ConnectionPool extends EventEmitter {
  private _host: string;
  private _port: number;
  private _database: string;
  private _username?: string;
  private _password?: string;
  private _tls: boolean;

  private _minConnections: number;
  private _maxConnections: number;
  private _maxIdleTime: number;
  private _maxLifetime: number;
  private _acquireTimeout: number;

  private _available: PooledConnection[] = [];
  private _inUse: Map<Connection, PooledConnection> = new Map();
  private _closed: boolean = false;

  // Statistics
  private _createdCount: number = 0;
  private _destroyedCount: number = 0;

  constructor(
    host: string,
    port: number,
    options: PoolOptions = {}
  ) {
    super();

    this._host = host;
    this._port = port;
    this._database = options.database || 'default';
    this._username = options.username;
    this._password = options.password;
    this._tls = options.tls !== false;

    this._minConnections = options.minConnections || 1;
    this._maxConnections = options.maxConnections || 10;
    this._maxIdleTime = options.maxIdleTime || 300000; // 5 minutes
    this._maxLifetime = options.maxLifetime || 3600000; // 1 hour
    this._acquireTimeout = options.acquireTimeout || 30000; // 30 seconds
  }

  /**
   * Get a connection from the pool
   */
  async getConnection(): Promise<PooledConnection> {
    if (this._closed) {
      throw new ConnectionError('Connection pool is closed');
    }

    // Try to get an available connection
    while (this._available.length > 0) {
      const pooledConn = this._available.pop()!;
      if (this._isConnectionValid(pooledConn)) {
        pooledConn.markUsed();
        this._inUse.set(pooledConn.connection, pooledConn);
        return pooledConn;
      } else {
        await this._destroyConnection(pooledConn);
      }
    }

    // Create new connection if under limit
    if (this._inUse.size < this._maxConnections) {
      const connection = new Connection(this._host, this._port, {
        database: this._database,
        username: this._username,
        password: this._password,
        tls: this._tls
      });

      await connection.connect();

      const pooledConn = new PooledConnection(connection, this);
      this._createdCount++;
      this._inUse.set(connection, pooledConn);
      return pooledConn;
    }

    // Wait for a connection to become available or timeout
    throw new ConnectionError(
      `Connection pool exhausted: ${this._inUse.size}/${this._maxConnections} connections in use`
    );
  }

  /**
   * Return a connection to the pool
   */
  async _returnConnection(pooledConn: PooledConnection): Promise<void> {
    if (this._inUse.has(pooledConn.connection)) {
      this._inUse.delete(pooledConn.connection);

      if (this._isConnectionValid(pooledConn)) {
        this._available.push(pooledConn);
      } else {
        await this._destroyConnection(pooledConn);
      }
    }
  }

  /**
   * Destroy a connection
   */
  private async _destroyConnection(pooledConn: PooledConnection): Promise<void> {
    try {
      await pooledConn.connection.close();
    } catch (error) {
      this.emit('error', error);
    }
    this._destroyedCount++;
  }

  /**
   * Check if a connection is valid
   */
  private _isConnectionValid(pooledConn: PooledConnection): boolean {
    return pooledConn.isValid(this._maxIdleTime, this._maxLifetime);
  }

  /**
   * Close the connection pool
   */
  async close(): Promise<void> {
    if (this._closed) {
      return;
    }

    this._closed = true;

    // Close all available connections
    const destroyPromises: Promise<void>[] = [];
    for (const pooledConn of this._available) {
      destroyPromises.push(this._destroyConnection(pooledConn));
    }

    // Close all in-use connections
    for (const pooledConn of this._inUse.values()) {
      destroyPromises.push(this._destroyConnection(pooledConn));
    }

    this._available = [];
    this._inUse.clear();

    await Promise.all(destroyPromises);
    this.emit('closed');
  }

  /**
   * Get pool statistics
   */
  get stats() {
    return {
      available: this._available.length,
      inUse: this._inUse.size,
      totalCreated: this._createdCount,
      totalDestroyed: this._destroyedCount,
      maxConnections: this._maxConnections
    };
  }

  /**
   * Check if pool is closed
   */
  get isClosed(): boolean {
    return this._closed;
  }

  /**
   * Get host
   */
  get host(): string {
    return this._host;
  }

  /**
   * Get port
   */
  get port(): number {
    return this._port;
  }

  /**
   * Get database
   */
  get database(): string {
    return this._database;
  }
}

/**
 * Create a connection pool
 */
export async function createPool(
  host: string,
  port: number,
  options: PoolOptions = {}
): Promise<ConnectionPool> {
  const pool = new ConnectionPool(host, port, options);
  // Optionally pre-warm the pool here
  return pool;
}
